### PR TITLE
[SPARK-17655][SQL]Remove unused variables declarations and definations in a WholeStageCodeGened stage

### PR DESCRIPTION
**Pull request by yaooqinn**
## What changes were proposed in this pull request?

A WholeStageCodeGened stage with multiple CodegenSupport Operators generates unused result rows and their associated buffer holders and row writers, which can be removed.
## How was this patch tested?

existing ut.

cc @davies @rxin @srowen

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -129,11 +131,17 @@ class CodegenContext {
...
+  private var mutableStates: mutable.ArrayBuffer[(String, String, String, Boolean)] =
+    mutable.ArrayBuffer.empty[(String, String, String, Boolean)]
 
-  def addMutableState(javaType: String, variableName: String, initCode: String): Unit = {
-    mutableStates += ((javaType, variableName, initCode))
+  def addMutableState(javaType: String, variableName: String,
+      initCode: String, isWholeStageResultVar: Boolean = false): Unit = {
+    if (isWholeStageResultVar) {
+      mutableStates = mutableStates.filterNot(state => (state._1 == javaType) && state._4)
+      mutableStates += ((javaType, variableName, initCode, isWholeStageResultVar))
```
Comment (rednaxelafx) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: So you're effectively letting later-added states to override earlier-added ones that are equivalent and replaceable. Could that be done in a add-if-absent way instead?
Also, what's the criteria for a state to be "replaceable" in your design? Could you please make that explicit as comment in the code?

**Review by rednaxelafx — COMMENTED** [review]
Hi, the goal of this PR looks interesting but the approach needs more explanation.
Could you please elaborate on the semantics of this new `isWholeStageResultVar` argument, give an example of how that affects the code in `GenerateUnsafeProjection`, and why is `GenerateUnsafeProjection.createCode()` the only place you're applying this new feature?

**Comment by kiszk** [comment]
Could you please add the simple generated code in the description before and after applying this PR?

**Comment by HyukjinKwon** [comment]
gentle ping @yaooqinn on ^.

**Pull request closed without merging**